### PR TITLE
Create and step env in a separate thread

### PR DESCRIPTION
**Pull request by witwolf**
#240 

social_bot envs (gazebo) requires `step` and `reset` run in the same thread which it created in ,
so add  a wrapper `ThreadPyEnvironment` that create and step it in a separate thread .

### CODE REVIEW

**Comment by hnyu** [comment]
> social_bot envs (gazebo) requires `step` and `reset` run in the same thread which it created in ,

Looking at the code, I suppose what you really mean is that `step` and `reset` have to be in a different thread than the main thread that creates the environment, right?

**Comment by emailweixu** [comment]
> #240
> 
> social_bot envs (gazebo) requires `step` and `reset` run in the same thread which it created in ,
> so add a wrapper `ThreadPyEnvironment` that create and step it in a separate thread .

Why?

**Comment by witwolf** [comment]
> > social_bot envs (gazebo) requires `step` and `reset` run in the same thread which it created in ,
> 
> Looking at the code, I suppose what you really mean is that `create`, `step` and `reset` have to be in a different thread than the main thread, right?

they can be in the main thread , only need to keep they are in the same thread

**Comment by witwolf** [comment]
> > #240
> > social_bot envs (gazebo) requires `step` and `reset` run in the same thread which it created in ,
> > so add a wrapper `ThreadPyEnvironment` that create and step it in a separate thread .
> 
> Why?

have not dig deep into gazebo yet,  just run a simple test with code below: 

```python
import social_bot
import gym
import threading

if __name__ == '__main__':
    env = gym.make('SocialBot-Pr2Gripper-v0')

    def _reset():
        env.reset()

    t = threading.Thread(target=_reset)
    t.start()
    t.join()
```

or 
```python
if __name__ == '__main__':
    env = gym.make('SocialBot-Pr2Gripper-v0')
    env.reset()

    def _step():
        env.step(env.action_space.sample())

    t = threading.Thread(target=_step)
    t.start()
    t.join()
```

it crashes at `reset`  (second code fragment crashes at `step`) with segmentation fault, and bt show it's related to `Ogre`

image

File: alf/environments/utils.py [hunk 1]
```diff
@@ -11,6 +11,12 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+
+# multiprocessing.dummy provides a pure *multithreaded* threadpool that works
+# in both python2 and python3 (concurrent.futures isn't available in python2).
```
Comment (emailweixu) [inline alf/environments/utils.py]: Can our code probably won't run using python2? I don't think so.

**Comment by witwolf** [comment]
> But it seems that all calls to the environment in the current play() are from the same thread (i.e. the main thread)

paly() is run in graph mode  , `step` is  called in thread managed by tf

File: alf/environments/utils.py (hunk 1 above)
Comment (witwolf) [reply alf/environments/utils.py]: removed (i copied these lines from `batched_py_environment.py`)

**Comment by emailweixu** [comment]
> > But it seems that all calls to the environment in the current play() are from the same thread (i.e. the main thread)
> 
> paly() is run in graph mode , `step` is called in thread managed by tf

I see.

**Pull request merged**